File: 22_专题/implicit_graph/RangeFinder/仅有删除且保证查询的元素存在/Finder-链表-严格O(1).go
package main

type Finder struct {
	n          int
	exist      []bool
	prev, next []int
}

// 建立一个包含0到n-1的集合.
func NewFinder(n int) *Finder {
	res := &Finder{
		n:     n,
		exist: make([]bool, n),
		prev:  make([]int, n),
		next:  make([]int, n),
	}
	for i := 0; i < n; i++ {
		res.exist[i] = true
		res.prev[i] = i - 1
		res.next[i] = i + 1
	}
	return res
}

// 0<=i<n.
func (fs *Finder) Has(i int) bool {
	return i >= 0 && i < fs.n && fs.exist[i]
}

// 0<=i<n.
func (fs *Finder) Erase(i int) bool {
	if !fs.Has(i) {
		return false
	}

	l, r := fs.prev[i], fs.next[i]
	if l >= 0 {
		fs.next[l] = r
	}
	if r < fs.n {
		fs.prev[r] = l
	}
	fs.exist[i] = false
	return true
}

// 返回`严格`小于i的最大元素,如果不存在,返回-1.
// !调用时需保证 Has(i)==true.
// 0<=i<n.
func (fs *Finder) Prev(i int) int {
	return fs.prev[i]
}

// 返回`严格`大于i的最小元素.如果不存在,返回n.
// !调用时需保证 Has(i)==true.
// 0<=i<n.
func (fs *Finder) Next(i int) int {
	return fs.next[i]
}

File: 22_专题/implicit_graph/RangeFinder/仅有删除且保证查询的元素存在/Finder-链表-严格O(1).ts
class Finder {
  private readonly _n: number
  private readonly _exist: Int8Array
  private readonly _prev: Int32Array
  private readonly _next: Int32Array

  constructor(n: number) {
    this._n = n
    this._exist = new Int8Array(n)
    this._prev = new Int32Array(n)
    this._next = new Int32Array(n)
    for (let i = 0; i < n; i++) {
      this._exist[i] = 1
      this._prev[i] = i - 1
      this._next[i] = i + 1
    }
  }

  has(i: number): boolean {
    return i >= 0 && i < this._n && !!this._exist[i]
  }

  erase(i: number): boolean {
    if (!this.has(i)) return false
    const l = this._prev[i]
    const r = this._next[i]
    if (l >= 0) this._next[l] = r
    if (r < this._n) this._prev[r] = l
    this._exist[i] = 0
    return true
  }

  /**
   * 返回`严格`小于i的最大元素,如果不存在,返回-1.
   * !调用时需保证i存在.
   */
  prev(i: number): number {
    return this._prev[i]
  }

  /**
   * 返回`严格`大于i的最小元素.如果不存在,返回n.
   * !调用时需保证i存在.
   */
  next(i: number): number {
    return this._next[i]
  }
}

export {}
